### ISSUE TITLE
file_known_and_unchanged is wasteful and should be converted to inodes

### ISSUE THREAD

**Issue by verygreen** [borgbackup/borg#909]
Moved from #908 
I guess another change is to turn cache.file_known_and_unchanged into just inode_known_and_unchanged.
Who cares what's the actual file name, since it's always stored in the archive anyway I believe?
Additional benefit, the "files_cache" huge file in .cache might become smaller as it no longer needs to hold path names.

This poses some difficulties for filesystems without inode numbers like vfat and on Windows, as mentioned by @ThomasWaldmann 

Windows (assuming they don't have inodes - I don't know much about ntfs) - some other unique identifier might be used, I guess. For filesystems that have no symlinks (like vfat) - file path might be unique enough. Still some races with content changing underneath, but at least nothing security critical (also hopefully nobody resurrects umsdos fs) (it's tempting to use filesize + create date (+modification date?) as unique id for vfat to handle renames more optimally there, but that might not be safe).
On nfs there's "file handle" that's unique (no idea if it's easily visible from userspace).
fuse and friends are harder, I guess.
ntfs has symlinks and hardlinks, so they must have have unique file ids similar to inode numbers of some sort.
These strategies could be determined by filesystem type (available from statfs(2) and could be called once we cross a mountpoint boundary assuming this is enabled, also separately checked for every path).
It should be noted that since inodes are not unique outside of a normal filesystem, really inode in the table should be some combination of actual inode number and filesystem id (similar to how nfs creates file handles). Filesystem id might be anything from mount path (unsafe) to underlying block device id (unsafe) to fs uuid (not bullet proof either) to some other options.

Using file path as generic fallback for inode number for cases when inode numbers are ignored for other reasons (commandline switch) is also possible, though it should be marked as "potentially suboptimal".

**Comment by ThomasWaldmann** [comment]
Currently, a mapping is used from key path_hash -> value FileCacheEntry (size, mtime, inode, age, chunk_ids).

key: path_hash = self.key.id_hash(safe_encode(os.path.join(self.cwd, path)))

So, the key hash could be just computed from different input data or get replaced by a tuple.

Algorithm:
```
if has_stable_inodes(path, st):
    fs_id = fs_uuid or blockdev_id or mount_path
    key = (fs_id, st.st_ino)
else:
    key = id_hash(path)  # fallback, see orig. code
```

**Comment by enkore** [comment]
A hash isn't even necessary - we can just pack the device number and the inode number in 16 bytes(!) each. But since we in fact don't even have the requirement of len(key) == 32 here, like we do for objects, we could just as well use `(dev, ino)` as the key directly.

**Comment by ThomasWaldmann** [comment]
Locally mounted fat, ntfs, iso, ext4: have non-null st_dev and st_ino.

http://www.mail-archive.com/bug-findutils@gnu.org/msg01212.html

**Comment by ThomasWaldmann** [comment]
Crap, os.statvfs() does not return f_type.

**Comment by enkore** [comment]
Yes I think the latter would be a better approach to this. Only some 'special' cases profit from this (but if they do, then quite a lot I suspect); normal hardlinked files in the same archive are already indexed by (dev, ino) anyway.

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/pull/1842#issuecomment-260184595 see there.

**Comment by enkore** [comment]
Re. compatibility with older Borg:

in ino-mode, truncate/leave files (this is valid). Store ino files cache as files.ino or so. No problems.

**Comment by ThomasWaldmann** [comment]
The code will dynamically switch between modes, so we need both in parallel.
Guess nothing bad will happen with old borg. The keys are bytes in both cases and they are both only computed in a one-way manner.

**Comment by ThomasWaldmann** [comment]
#1842 currently blocked by some C/Cython problem (access fsid) - can anybody help with that?

**Comment by ThomasWaldmann** [comment]
anyone?

**Comment by ThomasWaldmann** [comment]
Small related update:
- borg2 will archive the inode number of files now.
- And there is the idea in #8385.

**Comment by ThomasWaldmann** [comment]
About this ticket: does anybody know how to solve it?

The past tries have been unsuccessful - are there new ideas or new insights? Changing this before borg2 gets released would be a good time.
